### PR TITLE
[KOGITO-3222] Implement jpmml/trustypmml co-existence

### PR DESCRIPTION
**Pull request by gitgabrio**
This PR depends on https://github.com/kiegroup/kie-jpmml-integration/pull/40

@danielezonca @tarilabs  @jiripetrlik

see https://issues.redhat.com/browse/KOGITO-3222

This PR
At compile time:
1.  verify the presence of "jpmml" in the classpath
2. if jpmml is present, trusty-pmml compilaation is completely skipped

At runtime:
1. if "jpmml" is present, DMN will use it for pmml evaluation
2. if "jpmml" is present, no pmml endpoints will be generated (since trusty pmml compilation has completely been skipped)

### CODE REVIEW

**Comment by gitgabrio** [comment]
jenkins execute fdb

**Comment by gitgabrio** [comment]
jenkins please retest this

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 1]
```diff
@@ -88,15 +90,23 @@ public PredictionCodegen(List<PMMLResource> resources) {
         this.moduleGenerator = new PredictionContainerGenerator(applicationCanonicalName, resources);
     }
 
-    public static PredictionCodegen ofCollectedResources(Collection<CollectedResource> resources) {
+    public static PredictionCodegen ofCollectedResources(boolean isJPMMLAvailable,
+                                                         Collection<CollectedResource> resources) {
+        if (isJPMMLAvailable) {
+            return ofPredictions(Collections.emptyList());
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:96]: Can you please add a log message?

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 2]
```diff
@@ -88,15 +90,23 @@ public PredictionCodegen(List<PMMLResource> resources) {
...
+                .flatMap(r -> parsePredictions(r.basePath(),
+                                               Collections.singletonList(r.resource())).stream())
                 .collect(toList());
         return ofPredictions(dmnResources);
     }
 
-    public static PredictionCodegen ofJar(Path... jarPaths) throws IOException {
+    public static PredictionCodegen ofJar(boolean isJPMMLAvailable, Path... jarPaths) throws IOException {
+        if (isJPMMLAvailable) {
+            return ofPredictions(Collections.emptyList());
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:108]: Same as above

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 3]
```diff
@@ -119,7 +129,10 @@ public static PredictionCodegen ofJar(Path... jarPaths) throws IOException {
         return ofPredictions(pmmlResources);
     }
 
-    public static PredictionCodegen ofPath(Path... paths) throws IOException {
+    public static PredictionCodegen ofPath(boolean isJPMMLAvailable, Path... paths) throws IOException {
+        if (isJPMMLAvailable) {
+            return ofPredictions(Collections.emptyList());
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:134]: Same as above

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 4]
```diff
@@ -133,7 +146,10 @@ public static PredictionCodegen ofPath(Path... paths) throws IOException {
         return ofPredictions(resources);
     }
 
-    public static PredictionCodegen ofFiles(Path basePath, List<File> files) {
+    public static PredictionCodegen ofFiles(boolean isJPMMLAvailable, Path basePath, List<File> files) {
+        if (isJPMMLAvailable) {
+            return ofPredictions(Collections.emptyList());
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:151]: Same as above

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 1 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java]: @danielezonca 
done

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 2 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java]: @danielezonca 
Same as above

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 3 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java]: @danielezonca 
Same as above

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 4 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java]: @danielezonca 
Same as above

**Review by danielezonca — APPROVED** [review]

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 5]
```diff
@@ -88,15 +90,25 @@ public PredictionCodegen(List<PMMLResource> resources) {
...
                 .filter(r -> r.resource().getResourceType() == ResourceType.PMML)
-                .flatMap(r -> parsePredictions(r.basePath(), Collections.singletonList(r.resource())).stream())
+                .flatMap(r -> parsePredictions(r.basePath(),
+                                               Collections.singletonList(r.resource())).stream())
                 .collect(toList());
         return ofPredictions(dmnResources);
     }
 
-    public static PredictionCodegen ofJar(Path... jarPaths) throws IOException {
+    public static PredictionCodegen ofJar(boolean isJPMMLAvailable, Path... jarPaths) throws IOException {
```
Comment (evacchi) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:107]: I would advise merging https://github.com/kiegroup/kogito-runtimes/pull/741 before. Most of these factory methods have been deleted for good

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 6]
```diff
@@ -198,14 +217,15 @@ public PredictionContainerGenerator moduleGenerator() {
         if (resources.isEmpty()) {
             return Collections.emptyList();
         }
-        ModelBuilderImpl<KogitoPackageSources> modelBuilder = new ModelBuilderImpl<>( KogitoPackageSources::dumpSources,
-                new KnowledgeBuilderConfigurationImpl(getClass().getClassLoader()), new ReleaseIdImpl("dummy:dummy:0.0.0"), true, false );
+        ModelBuilderImpl<KogitoPackageSources> modelBuilder = new ModelBuilderImpl<>(KogitoPackageSources::dumpSources,
```
Comment (evacchi) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:220]: is the rest of this file just being reformatted or...?

File: kogito-build-parent/pom.xml [hunk 7]
```diff
@@ -129,7 +129,8 @@
     <version.org.keycloak>11.0.0</version.org.keycloak>
     <version.org.mockito>3.3.3</version.org.mockito>
     <version.org.mvel>2.4.7.Final</version.org.mvel>
-    <version.org.kie7>7.43.0.t20200824</version.org.kie7>
+<!--    <version.org.kie7>7.43.0.t20200824</version.org.kie7>-->
+    <version.org.kie7>7.43.0-SNAPSHOT</version.org.kie7>
```
Comment (evacchi) [inline kogito-build-parent/pom.xml:133]: let's merge after 7.43.0 is Final

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 8]
```diff
@@ -198,14 +217,15 @@ public PredictionContainerGenerator moduleGenerator() {
...
+                                                                                     new KnowledgeBuilderConfigurationImpl(getClass().getClassLoader()), new ReleaseIdImpl("dummy:dummy:0.0.0"), true, false);
         CompositeKnowledgeBuilder batch = modelBuilder.batch();
         for (PMMLResource resource : resources) {
             List<KiePMMLModel> kiepmmlModels = resource.getKiePmmlModels();
             for (KiePMMLModel model : kiepmmlModels) {
                 if (model.getName() == null || model.getName().isEmpty()) {
-                    String errorMessage = String.format("Model name should not be empty inside %s", resource.getModelPath());
+                    String errorMessage = String.format("Model name should not be empty inside %s",
```
Comment (evacchi) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:227]: all the remaining changes in this file are just whitespace changes right?

File: kogito-build-parent/pom.xml (hunk 7 above)
Comment (gitgabrio) [reply kogito-build-parent/pom.xml:133]: :+1:

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 8 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:227]: @evacchi 
Yes

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 6 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:220]: @evacchi 
Yup

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java (hunk 5 above)
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:107]: @evacchi 
:+1:

File: kogito-build-parent/pom.xml (hunk 7 above)
Comment (tarilabs) [reply kogito-build-parent/pom.xml:133]: Except for this line, thank you for keeping me in the loop Gabriele, lgtm

**Review by tarilabs — COMMENTED** [review]
https://github.com/kiegroup/kogito-runtimes/pull/737#discussion_r485017894

**Comment by gitgabrio** [comment]
jenkins please retest this

File: drools/kogito-pmml/src/test/java/org/kie/kogito/pmml/PmmlPredictionModelTest.java [hunk 9]
```diff
@@ -93,6 +94,7 @@ private static PMMLRuntime getPMMLRuntime() {
             public PMML4Result evaluate(String s, PMMLContext pmmlContext) {
                 return PMML_4_RESULT;
             }
+
```
Comment (jiripetrlik) [inline drools/kogito-pmml/src/test/java/org/kie/kogito/pmml/PmmlPredictionModelTest.java:97]: Why this new line?

**Review by jiripetrlik — APPROVED** [review]
One minor comment. Looks good to me.

File: drools/kogito-pmml/src/test/java/org/kie/kogito/pmml/PmmlPredictionModelTest.java (hunk 9 above)
Comment (gitgabrio) [reply drools/kogito-pmml/src/test/java/org/kie/kogito/pmml/PmmlPredictionModelTest.java]: @jiripetrlik 
Removed

**Pull request merged**